### PR TITLE
Fixed some moves' on-hit effects bypassing Substitutes where they shouldn't and some other things discovered along the way

### LINKED ISSUE #4519: Substitute does not block on-hit effects of several moves
### Description

Substitute does not block several secondary effects on moves.
For example Knock Off still knocks off items and Wake-Up Slap still wakes up sleeping Pokemon behind Substitutes.
Those examples might not be all of them, I'm working on testing more moves.
Moves bypassing Substitute where they shouldn't:

- Knock Off
- Wake-Up Slap
- Thousand Arrows
- Smack Down
- Salt Cure

Moves tested and works as expected: Trapping Moves, Throat Chop, Core Enforcer, Jaw Lock, Spirit Shackle, Incinerate

https://github.com/rh-hideout/pokeemerald-expansion/assets/149414898/f8e7751b-594f-4103-b187-5effcef7bdf8

### Version

1.8.3 (Latest release)

### Upcoming/master Version

_No response_

### Discord contact info

hedara

### PR DESCRIPTION
**Pull request by hedara90**
Fixed the effect of Knock Off, Salt Cure, Smack Down, Thousand Arrows, Wake-Up Slap and Smelling Salts not getting blocked by Substitutes where they should.
Fixed Sparkling Aria interaction with Shield Dust in Singles vs Doubles.
Fixed Wake-Up Slap and Smelling Salts getting boosted damage where they shouldn't vs Substitutes.

## Description

Encapsulated the calls to the effects of Knock Off, Salt Cure, Smack Down, Thousand Arrows, Wake-Up Slap and Smelling Salts in  `if (!DoesSubstituteBlockMove(...))` to check if they hit a Substitute or not.

Wrote a check for Sparkling Aria vs Shield Dust/Covert Cloak to implement the interaction where in doubles it can cure burns of supposedly protected Pokemon if Sparkling Aria hits more targets than the Shield Dust user. The check counts the number of Pokemon with more than 0 HP when it attempts to cure the Pokemon with Shield Dust, and if it's more than 2 it assumes that the condition to ignore Shield Dust has been met.

Encapsulated the damage multiplier for Wake-Up Slap and Smelling Salts in a `if (!DoesSubstituteBlockMove(...))`
 so that they don't get increased damage when hitting Substitutes.

Fixed broken doubles Shield Dust+Sparkling Aria test and extended it to cover the doubles case where all other Pokemon except Sparkling Aria user and Shield Dust user are fainted.
Wrote tests for all fixes and copied Shield Dust tests to Covert Cloak since they should always behave the same.

## Issue(s) that this PR fixes

Fixes #4519

## Feature(s) this PR does NOT handle:
Edge case of the Sparkling Aria burn curing implementation where if the Shield Dust Pokemon is the last one in the effect trigger order and at least one other valid target faints from the Sparkling Aria. Currently it will get cured of the burn, but it shouldn't.

## **Discord contact info**
hedara

### CODE REVIEW

**Comment by hedara90** [comment]
Fixed the PR failing on agbcc and some formatting pointed out by AlexOn1ine
Ready for review.

File: src/battle_util.c [hunk 1]
```diff
@@ -8615,8 +8615,12 @@ static inline u32 CalcMoveBasePower(u32 move, u32 battlerAtk, u32 battlerDef, u3
         break;
     case EFFECT_DOUBLE_POWER_ON_ARG_STATUS:
         // Comatose targets treated as if asleep
-        if ((gBattleMons[battlerDef].status1 | (STATUS1_SLEEP * (abilityDef == ABILITY_COMATOSE))) & gMovesInfo[move].argument)
+        if ((gBattleMons[battlerDef].status1 | (STATUS1_SLEEP * (abilityDef == ABILITY_COMATOSE))) & gMovesInfo[move].argument
+                && !((gMovesInfo[move].additionalEffects->moveEffect == MOVE_EFFECT_REMOVE_STATUS)
+                    && DoesSubstituteBlockMove(battlerAtk, battlerDef, move)))
```
Comment (AlexOn1ine) [inline src/battle_util.c:8620]:
```suggestion
        if ((gBattleMons[battlerDef].status1 | (STATUS1_SLEEP * (abilityDef == ABILITY_COMATOSE))) & gMovesInfo[move].argument
         && !((gMovesInfo[move].additionalEffects->moveEffect == MOVE_EFFECT_REMOVE_STATUS) && DoesSubstituteBlockMove(battlerAtk, battlerDef, move)))
```

Comment (AlexOn1ine) [inline src/battle_util.c:8620]:
```suggestion
        if ((gBattleMons[battlerDef].status1 | (STATUS1_SLEEP * (abilityDef == ABILITY_COMATOSE))) & gMovesInfo[move].argument
         && !((gMovesInfo[move].additionalEffects->moveEffect == MOVE_EFFECT_REMOVE_STATUS) && DoesSubstituteBlockMove(battlerAtk, battlerDef, move)))
```

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -5663,7 +5669,9 @@ static void Cmd_moveend(void)
                 }
                 break;
             case MOVE_EFFECT_REMOVE_STATUS: // Smelling salts, Wake-Up Slap, Sparkling Aria
-                if ((gBattleMons[gBattlerTarget].status1 & gMovesInfo[gCurrentMove].argument) && IsBattlerAlive(gBattlerTarget))
+                if ((gBattleMons[gBattlerTarget].status1 & gMovesInfo[gCurrentMove].argument)
+                        && IsBattlerAlive(gBattlerTarget)
+                        && !DoesSubstituteBlockMove(gBattlerAttacker, gBattlerTarget, gCurrentMove))
```
Comment (AlexOn1ine) [inline src/battle_script_commands.c:5674]:
```suggestion
                if ((gBattleMons[gBattlerTarget].status1 & gMovesInfo[gCurrentMove].argument)
                 && IsBattlerAlive(gBattlerTarget)
                 && !DoesSubstituteBlockMove(gBattlerAttacker, gBattlerTarget, gCurrentMove))
```

File: src/battle_script_commands.c [hunk 3]
```diff
@@ -5653,7 +5657,9 @@ static void Cmd_moveend(void)
                 }
                 break;
             case MOVE_EFFECT_SMACK_DOWN:
-                if (!IsBattlerGrounded(gBattlerTarget) && IsBattlerAlive(gBattlerTarget))
+                if (!IsBattlerGrounded(gBattlerTarget)
+                        && IsBattlerAlive(gBattlerTarget)
+                        && !DoesSubstituteBlockMove(gBattlerAttacker, gBattlerTarget, gCurrentMove))
```
Comment (AlexOn1ine) [inline src/battle_script_commands.c:5662]:
```suggestion
                if (!IsBattlerGrounded(gBattlerTarget)
                 && IsBattlerAlive(gBattlerTarget)
                 && !DoesSubstituteBlockMove(gBattlerAttacker, gBattlerTarget, gCurrentMove))
```

File: test/battle/ability/shield_dust.c [hunk 4]
```diff
@@ -121,19 +121,47 @@ SINGLE_BATTLE_TEST("Shield Dust does not block self-targeting effects, primary o
...
+            MESSAGE("Foe Vivillon's burn was healed.");
+            STATUS_ICON(opponentLeft, none: TRUE);
+        }
+        else
+        {
+            NONE_OF {
+                MESSAGE("Foe Vivillon's burn was healed.");
+                STATUS_ICON(opponentLeft, none: TRUE);
+            }
+        }
```
Comment (AlexOn1ine) [inline test/battle/ability/shield_dust.c:149]:
```suggestion
        if (moveToUse == MOVE_TACKLE) {
            MESSAGE("Foe Vivillon's burn was healed.");
            STATUS_ICON(opponentLeft, none: TRUE);
        } else {
            NONE_OF {
                MESSAGE("Foe Vivillon's burn was healed.");
                STATUS_ICON(opponentLeft, none: TRUE);
            }
        }
```

File: test/battle/item_effect/covert_cloak.c [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        ASSUME(MoveHasAdditionalEffectWithChance(MOVE_SMACK_DOWN, MOVE_EFFECT_SMACK_DOWN, 0) == TRUE);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_SKARMORY) { Item(ITEM_COVERT_CLOAK); }
+    } WHEN {
+        TURN { MOVE(player, move); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, move, player);
+        HP_BAR(opponent);
+        switch (move)
+        {
```
Comment (AlexOn1ine) [inline test/battle/item_effect/covert_cloak.c:65]:
```suggestion
        switch (move) {
```

File: test/battle/item_effect/covert_cloak.c [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    } WHEN {
+        TURN { MOVE(player, move); }
+        if (move == MOVE_METEOR_ASSAULT) {
+            TURN { SKIP_TURN(player); }
+        }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, move, player);
+        HP_BAR(opponent);
+        switch (move)
+        {
```
Comment (AlexOn1ine) [inline test/battle/item_effect/covert_cloak.c:111]:
```suggestion
        switch (move) {
```

File: test/battle/item_effect/covert_cloak.c [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            MESSAGE("Foe Wobbuffet's burn was healed.");
+            STATUS_ICON(opponentLeft, none: TRUE);
+        }
+        else
+        {
+            NONE_OF {
+                MESSAGE("Foe Wobbuffet's burn was healed.");
+                STATUS_ICON(opponentLeft, none: TRUE);
+            }
+        }
```
Comment (AlexOn1ine) [inline test/battle/item_effect/covert_cloak.c:149]:
```suggestion
        if (moveToUse == MOVE_TACKLE) {
            MESSAGE("Foe Wobbuffet's burn was healed.");
            STATUS_ICON(opponentLeft, none: TRUE);
        } else {
            NONE_OF {
                MESSAGE("Foe Wobbuffet's burn was healed.");
                STATUS_ICON(opponentLeft, none: TRUE);
            }
        }
```

File: test/battle/move_effect/smelling_salts.c [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                MESSAGE("Foe Seismitoad's SUBSTITUTE faded!"); // Smelling Salts does 86 damage, the sub has 122 HP, if hitting a sub it shouldn't get boosted damage.
+                MESSAGE("Foe Seismitoad was healed of paralysis!");
+                STATUS_ICON(opponent, none: TRUE);
+            }
+        }
+        else
+        {
+                MESSAGE("Foe Seismitoad was healed of paralysis!");
+                STATUS_ICON(opponent, none: TRUE);
```
Comment (AlexOn1ine) [inline test/battle/move_effect/smelling_salts.c:36]:
```suggestion
            MESSAGE("Foe Seismitoad was healed of paralysis!");
            STATUS_ICON(opponent, none: TRUE);
```

File: test/battle/move_effect/thousand_arrows.c [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        if (pokemon == SPECIES_SKARMORY)
+        {
+            MESSAGE("Foe Skarmory fell straight down!");
+            MESSAGE("Foe Skarmory used Celebrate!");
+        }
+        else
+        {
+            MESSAGE("Foe Scyther fell straight down!");
+            MESSAGE("Foe Scyther used Celebrate!");
+        }
```
Comment (AlexOn1ine) [inline test/battle/move_effect/thousand_arrows.c:44]:
```suggestion
        if (pokemon == SPECIES_SKARMORY) {
            MESSAGE("Foe Skarmory fell straight down!");
            MESSAGE("Foe Skarmory used Celebrate!");
        } else {
            MESSAGE("Foe Scyther fell straight down!");
            MESSAGE("Foe Scyther used Celebrate!");
        }
```

File: test/battle/move_effect/wake_up_slap.c [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                MESSAGE("Foe Seismitoad's SUBSTITUTE faded!"); // Smelling Salts does 86 damage, the sub has 122 HP, if hitting a sub it shouldn't get boosted damage.
+                MESSAGE("Foe Seismitoad woke up!");
+                STATUS_ICON(opponent, none: TRUE);
+            }
+        }
+        else
+        {
+                MESSAGE("Foe Seismitoad woke up!");
+                STATUS_ICON(opponent, none: TRUE);
```
Comment (AlexOn1ine) [inline test/battle/move_effect/wake_up_slap.c:36]:
```suggestion
            MESSAGE("Foe Seismitoad woke up!");
            STATUS_ICON(opponent, none: TRUE);
```

File: test/battle/move_effect/wake_up_slap.c [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_WAKE_UP_SLAP, player);
+        if (status1 == STATUS1_SLEEP)
+        {
+            MESSAGE("Foe Lotad fainted!");
+        }
+        else
+        {
+            NOT MESSAGE("Foe Lotad fainted!");
+            MESSAGE("Foe Lotad used Celebrate!");
+        }
```
Comment (AlexOn1ine) [inline test/battle/move_effect/wake_up_slap.c:61]:
```suggestion
        if (status1 == STATUS1_SLEEP) {
            MESSAGE("Foe Lotad fainted!");
        } else {
            NOT MESSAGE("Foe Lotad fainted!");
            MESSAGE("Foe Lotad used Celebrate!");
        }
```

File: test/battle/move_effect/wake_up_slap.c [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            MESSAGE("The SUBSTITUTE took damage for Foe Seismitoad!");
+            NONE_OF
+            {
+                MESSAGE("Foe Seismitoad's SUBSTITUTE faded!"); // Smelling Salts does 86 damage, the sub has 122 HP, if hitting a sub it shouldn't get boosted damage.
+                MESSAGE("Foe Seismitoad woke up!");
+                STATUS_ICON(opponent, none: TRUE);
+            }
+        }
+        else
+        {
```
Comment (AlexOn1ine) [inline test/battle/move_effect/wake_up_slap.c:34]:
```suggestion
        if (ability == ABILITY_INNER_FOCUS) {
            MESSAGE("The SUBSTITUTE took damage for Foe Seismitoad!");
            NONE_OF
            {
                MESSAGE("Foe Seismitoad's SUBSTITUTE faded!"); // Smelling Salts does 86 damage, the sub has 122 HP, if hitting a sub it shouldn't get boosted damage.
                MESSAGE("Foe Seismitoad woke up!");
                STATUS_ICON(opponent, none: TRUE);
            }
        } else {
```

**Review by AlexOn1ine — APPROVED** [review]
